src/sftp.rs

```diff
@@ -367,11 +370,32 @@ impl Sftp {
             Err(self.last_error())
         }
     }
+
+    /// Shuts the sftp connection down.
+    ///
+    /// This should not be called manually since the destructor takes care of it.
+    /// However, it is useful for async wrappers that take care of async destruction.
+    pub fn shutdown(&mut self) -> Result<(), Error> {
```

Comment:
I feel like this should be marked as `unsafe` as it can leave dangling references.
Alternatively, this can be made safe by storing the inner fields along these lines:

```rust
struct SftpInner {
    raw: *mut raw::LIBSSH2_SFTP,
    _sess: Arc<SessionInner>,
}

struct Sftp {
   inner: Option<SftpInner>,
}
```

and then arranging for your shutdown function to perform the shutdown and then `self.inner.take()` to release that state.

It's probably also a good idea to annotate it with `#[doc(hidden)]` to discourage its use outside of your async crate.

Reply:
The more I think about it the more I prefer the solution of marking shutdown as unsafe because:

1. There is no error that represents "already shutdown" in libssh2. Currently `Error` is a direct wrapper over errors that occur in libssh2 so (if I am not wrong) that would be either a breaking change or we reuse one of the error codes of libssh2.
2. We'd need to use the option every time we call a function for sftp.

What do you think?

Reply:
I think I'd prefer the `Option` approach; it feels more robust overall.
Regarding error codes, I think it's fine to use eg: `raw::LIBSSH2_ERROR_BAD_USE` in the case that we detect this

Reply:
Ok, I implemented that approach. Thanks for all the suggestions and looking over the code :-)